fms/modules/attention.py

```diff
@@ -583,6 +658,113 @@ def load_weights(
                     self.dense.bias, dense_bias, 1, [self.world_size], False
                 )
 
+    def load_qparams(
```

Comment:
I think we need to rethink the `load_weights` function so `load_qparams` can be a part of it. What I am thinking is that we have a list of modules like in qparams and we iterate over them. Then, each module implements a `tp_sharding` function that is dependent on the kind of linear function. In fact, given each of these linear modules comes from different places, the best option might be having these functions be separate and live in fms/modules/linear.py alongside get_linear(), and have the linear_config with the enum be passed in as well to decide which of the 3 (normal, GPTQ, FP8) to call.

The flow would then be something like:

```python
def load_weights(...):
    modules = ["qkv_fused", "dense"]
    sharding_dims = [0, 1]
    for module_name, sharding_dim in zip(modules, sharding_dims):
        fms.modules.linear.tp_sharding(module_name, sharding_dim, self.linear_config)
```

The internal function being called can do whatever needs to happen based on the assumption that the sharding dims being passed are those for the regular nn.Linear. We can discuss more offline as this is a big change

Reply:
I think @ani300 said is good and similar to vLLM's solution if I remember correctly. Maybe it should be called `load_shard`. This function will be quite complex to implement but maybe we can reuse some logic from vllm.